### PR TITLE
[quant][gpu][core] Implemented quantized add operator using cudnn [reland PR74463]

### PR DESCRIPTION
**Pull request by dzdang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74463

Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35218224](https://our.internmc.facebook.com/intern/diff/D35218224)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  // TODO: add shape checking? I think this is contingent on whether we support broadcasting in qadd, so maybe leave this out for now?
+  check_inputs(qa, qb);
+
+  // cudnn expects tensors to be at least 3D. So we will append dummy dimensions if the input tensors are not at least 3D
+  if (qa.dim() < 3) {
+    std::vector<int64_t> new_sizes{qa.sizes().vec()};
+    while (new_sizes.size() < 3) {
+      new_sizes.emplace_back(1);
+    }
+    qa = at::native::view(qa, new_sizes);
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:65]: nit: I think we should be using at::view, also we should be able to do `qa.view(....)` as well I think

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35009111).

File: aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  uint8_t input_b_alignment;
+  uint8_t output_alignment;
+  bool kReluFused;
+};
+std::unordered_map<CacheKey, cudnn_frontend::ManagedOpaqueDescriptor, at::native::ParamsHash<CacheKey>, at::native::ParamsEqual<CacheKey>> execution_plan_cache;
+}
+
+namespace at {
+namespace native {
+namespace {
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:34]: why do we have two anonymous namespaces? should we merge them?

File: test/quantization/core/test_quantized_op.py [hunk 3]
```diff
@@ -829,6 +829,44 @@ def test_qadd_relu_same_qparams(self):
             self.assertEqual(qCrelu_hat, qCrelu_out_hat,
                              msg="AddReLU.out failed")
 
+    """Tests the correctness of the cudnn add and add_relu op
+    (Similar to test_qadd_relu_different_qparams, will probably merge in the future)"""
+    @unittest.skipIf(not TEST_CUDNN, "cudnn is not enabled.")
+    @unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_op.py:835]: what is the effort to make this a flag accessible in pytorch, like TEST_CUDNN? I feel ideally we don't want to copy paste the code for every operator

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35009111).

File: aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+  // cudnn expects tensors to be at least 3D. So we will prepend dummy dimensions if the input tensors are not at least 3D
+  auto orig_sizes = qa.sizes().vec();
+  if (qa.dim() < 3) {
+    std::vector<int64_t> new_sizes{orig_sizes};
+    while (new_sizes.size() < 3) {
+      new_sizes.emplace_back(1);
+    }
+    // I think cudnn expects leading dimensions to be the dummy dimensions, so we need to swap
+    std::swap(new_sizes.front(), new_sizes.back());
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:71]: can we just use insert to do this?

Comment (jerryzh168) [reply aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:71]: oh, I think a cleaner way is to initialize a vector with the expected length and assign the last dim to orig_sizes

**Review by jerryzh168 — APPROVED** [review]
thanks for the comments for overall and individual ops, this makes it much easier to follow

File: aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+// We implement relu ( (a_int8 + b_int8 * ( b_scale/a_scale) ) ) * ( a_scale / out_scale )
+// which requires 4 cudnn ops (2 multiplication, 1 addition, and 1 relu ops)
+// Multiplication ops: rhs_mult_op, requant_op
+// Addition op: add_op
+// Relu op: relu_op
+template <bool kReluFused = false>
+Tensor add(Tensor qa, Tensor qb, double output_scale, int64_t output_zero_point) {
+  if (qa.numel() == 0) {
+    return Tensor{};
+  }
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:59]: should we check both qa and qb? and when does this happen?

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35009111).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35009111).

File: aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+  check_inputs(qa, qb);
+
+  // cudnn expects tensors to be at least 3D. So we will prepend dummy dimensions if the input tensors are not at least 3D
+  auto orig_sizes = qa.sizes().vec();
+  if (qa.dim() < 3) {
+    std::vector<int64_t> new_sizes(3, 1);
+    // cudnn expects leading dimensions to be the dummy dimensions
+    new_sizes.back() = qa.sizes().back();
+    if (qa.ndim() == 2) {
```
Comment (jjsjann123) [inline aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:71]: This line should be `qa.dim()` instead, similar to previous line 67.

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35218224).

**Pull request closed without merging**